### PR TITLE
#solve and permutation matrix from #factorize_lu

### PR DESCRIPTION
**Pull request by v0dro**
#factorize_lu would return a permutation array in a very strange manner, which left me confused until @andrewcsmith clarified it. So I've put a provision to return the permutation matrix along with the result.

Linear algebra equations can now be solved LAPACK free with the new #solve method.

### CODE REVIEW

**Comment by v0dro** [comment]
@mohawkjohn could you please check out this one too?

**Comment by v0dro** [comment]
@mohawkjohn Its done. With necessary changes.

**Comment by mohawkjohn** [comment]
@cjfuller I think this is ready for merge, wanna have a look and see if I missed anything?

**Comment by cjfuller** [comment]
Just a few nits (sorry, thanks for putting up with them!).  Looks good overall!  Can't say my linear algebra is fresh enough that I'm able to critically evaluate the solution, but that's why we have tests :)

**Comment by mohawkjohn** [comment]
Thanks, @cjfuller. Good catches.

**Comment by v0dro** [comment]
@cjfuller I've updated according to your comments.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -62,6 +62,15 @@ $ git checkout new-feature
 Before commiting any code, please read our
 [Contributor Agreement](http://github.com/SciRuby/sciruby/wiki/Contributor-Agreement).
 
+## C/C++ style guide
```
Comment (cjfuller) [inline CONTRIBUTING.md]: Thanks for this!

Comment (cjfuller) [inline CONTRIBUTING.md]: Though, now that I think of it, we have an (incredibly incomplete) style guide section in the wiki that might be easier to use than requiring a pull request every time: https://github.com/SciRuby/nmatrix/wiki/NMatrix-Developer-Guide#code-style

**Comment by cjfuller** [comment]
Looks great, thanks for the fixes!

File: CONTRIBUTING.md (hunk 1 above)
Comment (v0dro) [inline CONTRIBUTING.md]: I'll put that in there later.

**Comment by v0dro** [comment]
sure thing

File: spec/math_spec.rb [hunk 2]
```diff
@@ -343,4 +366,23 @@
       end
     end
   end
+
+  context "#solve", focus: true do
```
Comment (andrewcsmith) [inline spec/math_spec.rb]: The `focus: true` should be removed for general commit code.

Comment (v0dro) [inline spec/math_spec.rb]: I'll keep that in mind and fix that in a later commit.

Tired of rebasing now..

Comment (mohawkjohn) [inline spec/math_spec.rb]: @v0dro That's not really how it works. This is an open source project. If someone raises concerns about your code, we don't merge until we feel the concerns are satisfied.

I didn't realize that `:focus => true` does what it does, or I wouldn't have let this one slip through, either. It seems like it disables other tests? What's the point in having any tests at all if one of the tests disables all the others?

This is absolutely a fix that needs to be made before I will merge.

**Comment by v0dro** [comment]
@mohawkjohn I've updated it now. Could you merge?

**Comment by mohawkjohn** [comment]
I'd like to. It says I can't automatically, and my build is still borked. Any idea why it can't automatically merge? Is your master up to date?

**Comment by v0dro** [comment]
My master is up to date. This PR is quite old and based on an older master maybe thats why.

**Comment by mohawkjohn** [comment]
Can you rebase against the current master? If not, you're going to have to wait until I sort out my build environment, which isn't likely to be anytime soon. =/ Lots of traveling ahead of me.

**Comment by v0dro** [comment]
It will get changes from master which don't belong to me, if I have to rebase and you don't want the extra merge commit.

If the merge commit is fine with you then I don't mind.

**Comment by mohawkjohn** [comment]
Let's get @cjfuller to merge this one by hand now that he has access. That might be the easiest solution. Colin, is that okay with you?

**Comment by cjfuller** [comment]
Yep, I'll look at it after work tonight.

**Comment by cjfuller** [comment]
Rebased this manually onto master.  Closed by 29fd499f01bff3f0e61f321c69781e2a04cc27bb, 4241d241ca7744ca2ca5e090782588581160d42b, and 58d3a003bc2c34780532a68ba548b0ea92012ac7.

**Pull request closed without merging**